Playbooks: metrics (#5459)
mattermost/docs: source/playbooks/refining-and-improving.rst

Configuring retrospectives before a run
----------------------------------------

> Open Playbooks and select the **Playbooks** tab. Select the playbook and then select **Edit**. Select **Retrospective**. Move the toggle to **Enable Retrospective**.

Reviewer: It appears to take 6 "clicks" to configure a retro before a run. You've done a good job of documenting what a user specifically needs to do, but I wonder if there's an opportunity in the future to revisit and streamline the process overall through a UX redesign.

Author: Definitely worth streamlining it. I know there are ongoing efforts to refactor parts of playbooks. I'm happy to raise this with the team. :)

> Open Playbooks and select the **Playbooks** tab. Locate the playbook you want to modify, select the **...** icon under **Actions** and then select **Edit**. In the next screen, select **Retrospective**. Move the toggle to **Enable Retrospective**.

Author: @cwarnermm, I've added a flow with one less click to it. I don't have a proposal in mind around how to improve the flow but I know refactoring work is in progress so I'll keep an eye on that and see where we can improve this and other flows.

Reviewer: This is a great update, @justinegeffen!

.. image:: ../images/Retro.gif
   :alt: Create and publish retrospective reports.

Metrics
~~~~~~~

> Use metrics to identify key areas in which you want to measure performance and improvement, and extract valuable insights. 

Reviewer: ```suggestion
Use metrics to identify key areas where you want to extract valuable insights by measuring performance and improvement. 
```

> These metrics can be anything that's of interest to you and your team. For example, for a software release playbook you might want to have a metric tracking how many bugs were detected during a run. The output of the metrics you've added is provided in the retrospective and, over time, you can use the data to examine anomalies and refine goals.

Reviewer: ```suggestion
These metrics can be anything that's of interest to you and your team. For example, for a software release playbook, you might want to have a metric tracking how many bugs were detected during a run. The output of the metrics you've added is provided in the retrospective and, over time, you can use the data to examine anomalies and refine goals.
```

Reviewer: If you decide to move the dependency information to the top of this section, perhaps reword the last line here to: "Over time, you can use metrics across retrospectives to examine anomalies and refine goals."

> Another example is a support incident playbook. A metric such as time to resolution can be applied and used to identify areas that need more refinement such as tasks that might work better if they're split up so the goal is reached faster.

Reviewer: ```suggestion
Another example is a support incident playbook. The time to resolution metric can be applied and used to identify areas that need more refinement, such as tasks that might work better if they're split up so goals are reached faster.
```

> Metrics are enabled when you enable retrospectives. Calibrate the type of metric you want to measure, and once a run is finished you can view the output in the retrospective. You can have multiple metrics configured per playbook and you can edit them at any time.

Author: @gbochora, could you tell me a bit more about what happens if metrics are deleted from a playbook? So if I have two types of metrics configured and I decide to remove one, does that impact anything going forward?

Reviewer 2: - Key metrics dashboard(Playbook preview -> Key metrics tab) will not show information for removed metrics anymore. 
- The corresponding metric field will be removed from the retrospective form. This will happen for published retrospectives too.

> Metrics are enabled when you enable retrospectives. Calibrate the type of metric you want to measure, and once a run is finished you can view the output in the retrospective. You can have multiple metrics configured per playbook and you can edit them at any time.

Reviewer: "Metrics are enabled when you enable retrospectives." -- I feel that this may be important information to include at the beginning of this subsection because it's communicating a dependency between retros & metrics that the user may otherwise be unaware of.

Reviewer: In addition, if this content was moved to the beginning of this section, you'd be able to remove the redundant text above: "The output of the metrics you've added is provided in the retrospective"

> When you delete configured metrics from a playbook, you will no longer see those metrics in the dashboard. In addition, the corresponding metric field will be removed from the retrospective form and from published retrospectives.

Author: @gbochora, one more question - I'd imagine that if someone inadvertently deleted a metric and it disappeared from all retros, they may have a problem, especially if it's a valuable one.

Should we encourage/suggest that retros are downloaded/copied so data doesn't get lost? I don't want to set users up for failure/data loss. :)

Reviewer 2: Data will not be lost, it will still be stored in the database. Currently, there is no way to export it, but we are planning to add such functionality as far as I know.
Also, one question, should we add more information about the playbook metrics dashboard?

Author: Great suggestion; do you have anything specific in mind or any specs I can look at?

Reviewer 2: This is from Epic description(just updated few details):
For each run, participants can enter an actual value per metric under the retrospective section. They can edit it as many times as they'd like until publishing the retrospective at which point both the metrics and report will no longer be editable. The final value only shows up in the playbook dashboard once retrospective has been published.

The playbook dashboard reports on each metric for all runs by showing:

- average value for all runs
- last 10-run average value and difference with prev 10-run average
- value range for all runs
- target value
- actual values chart for last 10-runs

The lower section shows a list of finished runs with metrics values. Customers can filter or sort a list by metrics values.

Please let me know if you have any questions.

Design:
![image](https://user-images.githubusercontent.com/4368372/157894280-1a757f96-fd8e-42b9-9f0d-314dd97473a9.png)

> When you delete configured metrics from a playbook, you will no longer see those metrics in the dashboard. In addition, the corresponding metric field will be removed from the retrospective form and from published retrospectives.

Author: ```suggestion
When you delete configured metrics from a playbook, the data is not delete but you will no longer see those metrics in the dashboard. In addition, the corresponding metric field will be removed from the retrospective form and from published retrospectives.
```

These metrics can be anything that's of interest to you and your team. For example, for a software release playbook you might want to have a metric tracking how many bugs were detected during a run. The output of the metrics you've added is provided in the retrospective. Over time, you can use metrics across retrospectives to examine anomalies and refine goals.

Another example is a support incident playbook. The time to resolution metric can be applied and used to identify areas that need more refinement, such as tasks that might work better if they're split up so goals are reached faster.

When you delete configured metrics from a playbook, the data is not delete but you will no longer see those metrics in the dashboard. In addition, the corresponding metric field will be removed from the retrospective form and from published retrospectives.

Metrics dashboard
^^^^^^^^^^^^^^^^^

> For each run, you can enter an actual value per metric under the retrospective section. You can edit it any many times as you'd like until you've published the retrospective at which point both the metrics and report will no longer be editable. The final value only shows up in the playbook dashboard once the retrospective has been published.

Author: ```suggestion
For each run, you can enter a value per metric under the retrospective section. This value can be based on whatever you are measuring, such as number of bugs found per run. 

You can edit this value as many times as you'd like - for example to ensure you have the correct parameters for your metric - until you've published the retrospective at which point both the metrics and report will no longer be editable. The final value only shows up in the playbook dashboard once the retrospective has been published.
```

.. image:: ../images/Retro.gif
   :alt: Create and publish retrospective reports.

Metrics
~~~~~~~

Use metrics to identify key areas where you want to extract valuable insights by measuring performance and improvement. Metrics are enabled when you enable retrospectives. Calibrate the type of metric you want to measure, and once a run is finished you can view the output in the retrospective. You can have multiple metrics configured per playbook and you can edit them at any time. Metrics can be configured based on numeric, time, or value input.

These metrics can be anything that's of interest to you and your team. For example, for a software release playbook you might want to have a metric tracking how many bugs were detected during a run. The output of the metrics you've added is provided in the retrospective. Over time, you can use metrics across retrospectives to examine anomalies and refine goals.

Another example is a support incident playbook. The time to resolution metric can be applied and used to identify areas that need more refinement, such as tasks that might work better if they're split up so goals are reached faster.

> When you delete configured metrics from a playbook, the data is not delete but you will no longer see those metrics in the dashboard. In addition, the corresponding metric field will be removed from the retrospective form and from published retrospectives.

Reviewer: ```suggestion
When you delete configured metrics from a playbook, the data is not deleted but you no longer see those metrics in the dashboard. In addition, the corresponding metric field is removed from the retrospective form and from published retrospectives.
```

Metrics dashboard
^^^^^^^^^^^^^^^^^

> For each run, you can enter a value per metric under the retrospective section. This value can be based on whatever you are measuring, such as number of bugs found per run. 

Author: @gbochora, I know you've approved this PR (thank you!) - if you'd be open to reviewing this section I'd appreciate it.

Reviewer 2: Looks nice!

> You can edit this value as many times as you'd like - for example to ensure you have the correct parameters for your metric - until you've published the retrospective at which point both the metrics and report will no longer be editable. The final value only shows up in the playbook dashboard once the retrospective has been published.

Reviewer: ```suggestion
You can edit this value as many times as you'd like - for example to ensure you have the correct parameters for your metric - until you've published the retrospective at which point both the metrics and report are no longer editable. The final value only shows up in the playbook dashboard once the retrospective has been published.
```
